presto-main/src/main/java/com/facebook/presto/type/khyperloglog/KHyperLogLog.java

```diff
@@ -0,0 +1,285 @@
+    {
+        return SIZE_OF_KHYPERLOGLOG +
+                SIZE_OF_RBTREEMAP +
+                minhash.size() * SIZE_OF_LONG +
+                minhash.values().stream().mapToInt(HyperLogLog::estimatedInMemorySize).sum();
+    }
+
+    public int estimatedSerializedSize()
+    {
+        return SIZE_OF_BYTE +
+                3 * SIZE_OF_INT +
+                minhash.size() * (SIZE_OF_LONG + SIZE_OF_INT) +
+                minhash.values().stream().mapToInt(HyperLogLog::estimatedSerializedSize).sum();
+    }
+
+    public void add(long value, long uii)
+    {
+        update(Murmur3Hash128.hash64(value), uii);
+    }
+
+    public void add(Slice value, long uii)
+    {
+        update(Murmur3Hash128.hash64(value), uii);
+    }
+
+    private void update(long hash, long uii)
+    {
+        if (minhash.containsKey(hash)) {
+            minhash.get(hash).add(uii);
+        }
+        else if (isExact() || hash < minhash.lastLongKey()) {
+            HyperLogLog hll = HyperLogLog.newInstance(hllBuckets);
+            hll.add(uii);
+            minhash.put(hash, hll);
+            removeOverflowEntries();
+        }
+    }
+
+    public long cardinality()
+    {
+        if (isExact()) {
+            return minhash.size();
+        }
+
+        // Intuition is: get the stored hashes' density, and extrapolate to the whole Hash output range.
+        // Since Hash output range (2^64) cannot be stored in long type, I use half of the range
+        // via Long.MAX_VALUE and also divide the hash values' density by 2. The "-1" is bias correction
+        // detailed in "On Synopses for Distinct-Value Estimation Under Multiset Operations" by Beyer et. al.
+        long hashesRange = minhash.lastLongKey() - Long.MIN_VALUE;
+        double halfDensity = Long.divideUnsigned(hashesRange, minhash.size()) / 2D;
+        return (long) (HASH_OUTPUT_HALF_RANGE / halfDensity);
+    }
+
+    public KHyperLogLog mergeWith(KHyperLogLog other)
+    {
+        LongBidirectionalIterator iterator = other.minhash.keySet().iterator();
+        while (iterator.hasNext()) {
+            long key = iterator.nextLong();
+            if (minhash.containsKey(key)) {
+                minhash.get(key).mergeWith(other.minhash.get(key));
+            }
+            else {
+                minhash.put(key, other.minhash.get(key));
+            }
+        }
+        removeOverflowEntries();
+        return this;
+    }
+
+    public double reidentificationPotential(long threshold)
+    {
+        int highlyUniqueValues = 0;
+        for (HyperLogLog hll : minhash.values()) {
+            if (hll.cardinality() <= threshold) {
+                highlyUniqueValues++;
+            }
+        }
+        return (double) highlyUniqueValues / minhash.size();
+    }
+
+    public Map<Long, Double> uniquenessDistribution()
+    {
+        return uniquenessDistribution(minhash.size());
+    }
+
+    public Map<Long, Double> uniquenessDistribution(long histogramSize)
```

Comment:
Let's find a way to do this without boxed types

Reply:
Using fastutil Long2DoubleMap